pkg/resources/provisioning.cattle.io/v1/cluster/validator.go

```diff
@@ -113,6 +131,83 @@ func (p *provisioningAdmitter) Admit(request *admission.Request) (*admissionv1.A
+	if newCluster.Spec.RKEConfig == nil {
+		return admission.ResponseAllowed()
+	}
+	if request.Operation == admissionv1.Create {
+		if slices.ContainsFunc(newCluster.Spec.AgentEnvVars, func(envVar rkev1.EnvVar) bool {
+			return envVar.Name == systemAgentVarDirEnvVar
+		}) {
+			return admission.ResponseBadRequest(
+				fmt.Sprintf(`"%s" cannot be set within "cluster.Spec.RKEConfig.AgentEnvVars": use "cluster.Spec.RKEConfig.DataDirectories.SystemAgent"`, systemAgentVarDirEnvVar))
+		}
+		return admission.ResponseAllowed()
+	}
+	if request.Operation == admissionv1.Update {
+		var oldCount, newCount int
+		for _, envVar := range oldCluster.Spec.AgentEnvVars {
+			if envVar.Name == systemAgentVarDirEnvVar {
+				oldCount++
+			}
+		}
+		for _, envVar := range newCluster.Spec.AgentEnvVars {
+			if envVar.Name == systemAgentVarDirEnvVar {
+				newCount++
+			}
+		}
+		if newCount > oldCount {
+			return admission.ResponseBadRequest(
+				fmt.Sprintf(`"%s" cannot be set within "cluster.Spec.RKEConfig.AgentEnvVars": use "cluster.Spec.RKEConfig.DataDirectories.SystemAgent"`, systemAgentVarDirEnvVar))
+		}
+	}
+
+	return admission.ResponseAllowed()
+}
+
+// validateDataDirectories will validate updates to the cluster object to ensure the data directories are not changed.
+// The only exception when a data directory is allowed to be changed is if cluster.Spec.AgentEnvVars has an env var with
+// a name of "CATTLE_AGENT_VAR_DIR", which Rancher will perform a one-time migration to set the
+// cluster.Spec.RKEConfig.DataDirectories.SystemAgent field for the cluster. validateAgentEnvVars will ensure
+// "CATTLE_AGENT_VAR_DIR" is not added, so this exception only applies to the one-time Rancher migration.
+func (p *provisioningAdmitter) validateDataDirectories(request *admission.Request, oldCluster, newCluster *v1.Cluster) *admissionv1.AdmissionResponse {
+	if request.Operation != admissionv1.Update {
+		return admission.ResponseAllowed()
+	}
+	if newCluster.Spec.RKEConfig == nil {
+		return admission.ResponseAllowed()
+	}
+	oldSystemAgentVarDirEnvVars := slices.DeleteFunc(slices.Clone(oldCluster.Spec.AgentEnvVars), func(envVar rkev1.EnvVar) bool {
+		return envVar.Name != systemAgentVarDirEnvVar
+	})
+	// will be true up until rancher performs one time migration to set SystemAgent data directory if env var was previously set
+	if len(oldSystemAgentVarDirEnvVars) > 0 {
+		newSystemAgentVarDirEnvVars := slices.DeleteFunc(slices.Clone(newCluster.Spec.AgentEnvVars), func(envVar rkev1.EnvVar) bool {
+			return envVar.Name != systemAgentVarDirEnvVar
+		})
+		if !reflect.DeepEqual(newSystemAgentVarDirEnvVars, oldSystemAgentVarDirEnvVars) || oldCluster.Spec.RKEConfig.DataDirectories.SystemAgent != newCluster.Spec.RKEConfig.DataDirectories.SystemAgent {
+			if newCluster.Spec.RKEConfig.DataDirectories.SystemAgent != oldSystemAgentVarDirEnvVars[len(oldSystemAgentVarDirEnvVars)-1].Value {
```

Comment:
Nit: I think that this is a bit tough to follow. I think this would also be easier to write once the actual env var is extracted and you can move out of slices.

In addition, since the second if statement doesn't have an else (or any other additional logic). it can be combined into one if statement.